PointCloud3 publication from ORB_SLAM3

shandilya1998 opened the pull request:
Problem
=======
1. OctoMap needs sensor origin position in global map reference to build octomap from point clouds
2. Not sufficient matches detect to build a sufficiently dense map for planning effectively
3. Point clouds from tracking thread not needed for octomap building

Solution
========
1. Added octomap dependencies in CMakeLists.txt and package.xml
2. Added custom_interfaces::msg::PointCloud3 dependencies in CMakeLists.txt and package.xml
3. Updated all point cloud publication related method definitions to use custom_interfaces::msg::PointCloud3
4. Increased the number of features from 1250 to 3600 in tello.yaml
5. Removed all instances of `FrameMapPointUpdateCallback(mTrackedPoints, tcw)` from System.cc
6. Implemented ORB_SLAM3 to ROS coordinate system transformation for camera position

Note
====
1. Added `SophusToGeometryMsgTransform(...)` in MonocularSlamNode. The method may be used in other nodes as well
2. Point Clouds are only published when the local mapping thread updates the global map, thus ensuring cosistency between the octomap and ORB_SLAM3 map

shandilya1998 committed:
PointCloud3 publication from ORB_SLAM3

Problem
=======
1. OctoMap needs sensor origin position in global map reference to build
   octomap from point clouds
2. Not sufficient matches detect to build a sufficiently dense map for
   planning effectively
3. Point clouds from tracking thread not needed for octomap building

Solution
========
1. Added octomap dependencies in CMakeLists.txt and package.xml
2. Added custom_interfaces::msg::PointCloud3 dependencies in
   CMakeLists.txt and package.xml
3. Updated all point cloud publication related method definitions to use
   custom_interfaces::msg::PointCloud3
4. Increased the number of features from 1250 to 3600 in tello.yaml
5. Removed all instances of `FrameMapPointUpdateCallback(mTrackedPoints,
   tcw)` from System.cc
6. Implemented ORB_SLAM3 to ROS coordinate system transformation for
   camera position

Note
====
1. Added `SophusToGeometryMsgTransform(...)` in MonocularSlamNode.
   The method may be used in other nodes as well

vshkar commented on src/monocular.cpp:
```diff
@@ -11,6 +11,9 @@ MonocularSlamNode::MonocularSlamNode() : SlamNode("orbslam3_mono_node"){
 	mpSlamStartupService = this->create_service<StartupSlam>("~/start_slam", std::bind(&MonocularSlamNode::InitialiseSlamNode, this, std::placeholders::_1, std::placeholders::_2));
+	const double data[] = {0, 0, 1, -1, 0, 0, 0, -1, 0};
+	Eigen::Matrix3d orbToROSTransform(data);
+	mpOrbToROSTransform = orbToROSTransform;
```
why create a orbToRosTransform if its being later assigned to mpOrbToRosTransform?

shandilya1998 replied:
I think I wrote this code separately to test it out and then used it here. Will correct the same.

vshkar commented on src/monocular.cpp:
```diff
@@ -11,6 +11,9 @@ MonocularSlamNode::MonocularSlamNode() : SlamNode("orbslam3_mono_node"){
 	RCLCPP_INFO(this->get_logger(), "Created SLAM Object for MORBSLAM in Mono Slam Node");
 #endif
 	mpSlamStartupService = this->create_service<StartupSlam>("~/start_slam", std::bind(&MonocularSlamNode::InitialiseSlamNode, this, std::placeholders::_1, std::placeholders::_2));
+	const double data[] = {0, 0, 1, -1, 0, 0, 0, -1, 0};
```
is this a rotation matrix? can you please name the data variable more aptly for the code readers? Do you think we can use the `>>` Eigen::Matrix operator instead of assigning to data then using the constructor?

shandilya1998 replied:
Updated the same.

vshkar commented on src/monocular.cpp:
```diff
@@ -196,12 +203,28 @@ MapMsg MonocularSlamNode::MapPointsToPointCloud (std::vector<ORB_SLAM3::MapPoint
             data_array[1] = map_pt[1];
             data_array[2] = map_pt[2];
-
-            memcpy(cloud_data_ptr + (i * cloud.point_step), data_array, num_channels * sizeof(float));
```
can you help me understand where the memory for the point cloud is being allocated?

shandilya1998 replied:
You can find the memory allocation here: https://github.com/shandilya1998/ros2_orb_slam3/blob/b20237b9d8d3e0e92c8fce291865567fdbe37117/src/monocular.cpp#L187

vshkar replied:
cloud itself is unassigned right? 
`MapMsg::UniquePtr msg = std::make_unique<MapMsg>();`

shandilya1998 replied:
memory is allocated to the cloud, but it is empty. I wanted to point to the data array without having to use the complex notation.

shandilya1998 committed:
Slam Pose Inversion and Global Map Point Publication

Problem
=======
1. OctoMap needs transformation from camera to world, slam publishes
   world to camera
2. OctoMap needs global map points for map update
3. Need to convert from ORBSLAM coordinate system to ROS2 coordinate
   system for downstream tasks

Solution
========
1. Inverted pose transform published to be from camera to world frame
2. Removed map point transformation from world to camera frame before
   publication
3. Updated mpOrbToROS2Transform to describe rotation along x, previously
   it described rotation along y axis

Note
====
1. Need to decide on convention for point cloud publication to extend to
   other slam and 3D sensing sensors
2. OctoMap currently build is rotated. Need to fix the same

shandilya1998 committed:
PR review comment resolution

Problem
=======
1. Use `<<` instead of explicitly creating and assigning an array

Solution
========
1. Updated src/monocular.cpp to update initialisation of
   mpOrbtoROSTransform

Note
====
1. Tested with tello drone. RViz visualisation corresponds to 3D map of
   environment after z axis inversion
